omod/src/main/java/org/openmrs/module/owa/web/controller/OwaRestController.java

```diff
@@ -151,77 +157,110 @@
 					} else {
 						ZipEntry entry = zip.getEntry("manifest.webapp");
 						if (entry == null) {
 							message = messageSourceService.getMessage("owa.manifest_not_found");
 							log.warn("Manifest file could not be found in app");
-							file.delete();
 							session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-							response.sendError(500, message);
+							response.sendError(400, message);
+							return new ResponseEntity<List<App>>(HttpStatus.BAD_REQUEST);
 						} else {
 							String contextPath = request.getScheme() + "://" + request.getServerName() + ":"
 									+ request.getServerPort() + request.getContextPath();
 							appManager.installApp(file, fileName, contextPath);
-							file.delete();
 							message = messageSourceService.getMessage("owa.app_installed");
 							response.setStatus(200);
+							file.delete();
 							session.setAttribute(WebConstants.OPENMRS_MSG_ATTR, message);
+							appManager.reloadApps();
+							appList = appManager.getApps();
+							return new ResponseEntity<List<App>>(appList, HttpStatus.OK);
 						}
 					}
-				}
-				catch (Exception e) {
+				}catch(Exception e){
 					message = messageSourceService.getMessage("owa.not_a_zip");
 					log.warn("App is not a zip archive");
 					file.delete();
 					session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-					response.sendError(500, message);
+					response.sendError(400, message);
+					return new ResponseEntity<List<App>>(HttpStatus.BAD_REQUEST);
 				}
-			} else {
-				message = messageSourceService.getMessage("owa.invalid_url");
-				log.warn("Invalid url");
-				session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
-				response.sendError(400, message);
 			}
-			appManager.reloadApps();
-			appList = appManager.getApps();
+			catch (FileNotFoundException e) {
+				e.printStackTrace();
+				log.warn(e.getMessage());
+				session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, e.getMessage());
+				response.sendError(400, e.getMessage());
+				return new ResponseEntity<List<App>>(HttpStatus.BAD_REQUEST);
+			}
 		}
-		return appList;
+		return new ResponseEntity<List<App>>(HttpStatus.FORBIDDEN);
 	}
 
 	@RequestMapping(value = "/rest/owa/allowModuleWebUpload", method = RequestMethod.GET)
 	@ResponseBody
-	public boolean allowWebAdmin(HttpServletRequest request, HttpServletResponse response) {
-		 return ModuleUtil.allowAdmin();
+	public Boolean allowWebAdmin(HttpServletRequest request, HttpServletResponse response) {
+		return ModuleUtil.allowAdmin();
+	}
+
+	private String getPassedFileName(String installUrl) {
+		String passedFileName = null;
+		if (installUrl.contains("file_path=")) {
+			passedFileName = StringUtils.substringBetween(installUrl, "file_path=", ".zip");
+		} else {
+			passedFileName = FilenameUtils.getName(installUrl);
+		}
+		return getFileName(passedFileName);
+	}
+
+	private String getFileName(String passedFileName) {
```

Comment:
What is this method doing to the passed file name? I do not see it reflected in the method name.
Can you also create some unit tests for this method?